### PR TITLE
core(tsc): add type checking to fonts gatherer

### PR DESCRIPTION
**Pull request by brendankenny**
one of the big @ts-nochecked gatherers

### CODE REVIEW

File: lighthouse-core/gather/gatherers/fonts.js [hunk 1]
```diff
@@ -138,27 +172,42 @@ class Fonts extends Gatherer {
...
-    return Promise.all(
+    /** @type {Promise<[Array<LH.Artifacts.Font>, Array<LH.Artifacts.Font|FontGatherError>]>} */
+    const fontData = Promise.all(
       [
         driver.evaluateAsync(`(${getAllLoadedFonts.toString()})(${args})`),
         driver.evaluateAsync(`(${getFontFaceFromStylesheets.toString()})()`),
       ]
-    ).then(([loadedFonts, fontFaces]) => {
-      return loadedFonts.map(fontFace => {
-        if (fontFace.err) {
```
Comment (brendankenny) [inline lighthouse-core/gather/gatherers/fonts.js]: as far as I can tell, this was a bug. `fontFace` came from `getAllLoadedFonts` while it was `getFontFaceFromStylesheets` that generates the error objects it's looking for

File: lighthouse-core/gather/gatherers/fonts.js [hunk 2]
```diff
@@ -44,39 +57,49 @@ function getAllLoadedFonts(descriptors) {
...
           const fontsObject = {
+            // @ts-ignore (currently) non-standard Chrome extension to CSSStyleDeclaration
             display: rule.style.fontDisplay || 'auto',
-            family: rule.style.fontFamily.replace(/"|'/g, ''),
+            family: rule.style.fontFamily ? rule.style.fontFamily.replace(/"|'/g, '') : '',
             stretch: rule.style.fontStretch || 'normal',
             style: rule.style.fontStyle || 'normal',
             weight: rule.style.fontWeight || 'normal',
             variant: rule.style.fontVariant || 'normal',
+            // @ts-ignore (currently) non-standard Chrome extension to CSSStyleDeclaration
```
Comment (paulirish) [inline lighthouse-core/gather/gatherers/fonts.js]: i dont think these are non-standard, they're just fairly new.

File: lighthouse-core/gather/gatherers/fonts.js [hunk 3]
```diff
@@ -92,44 +115,55 @@ function getFontFaceFromStylesheets() {
...
+        data.push(...getSheetsFontFaces(cssStylesheet));
       }
     } catch (err) {
-      promises.push({err: {message: err.message, stack: err.stack}});
+      data.push({err: {message: err.message, stack: err.stack}});
     }
   }
   // Flatten results
-  return Promise.all(promises).then(fontFaces => [].concat(...fontFaces));
+  return Promise.all(corsDataPromises).then(corsFontFaces => data.concat(...corsFontFaces));
```
Comment (paulirish) [inline lighthouse-core/gather/gatherers/fonts.js]: nice clean up separating promises and data!

File: typings/externs.d.ts [hunk 4]
```diff
@@ -15,6 +15,13 @@ declare global {
     code?: string;
   }
 
+  /** Make properties K in T optional. */
+  type MakeOptional<T, K extends keyof T> = {
```
Comment (patrickhulce) [inline typings/externs.d.ts]: Does this actually get used anywhere? Seems like you might have opted for `Exclude` instead

File: lighthouse-core/gather/gatherers/fonts.js (hunk 2 above)
Comment (brendankenny) [reply lighthouse-core/gather/gatherers/fonts.js]: adding link to that discussion

**Pull request merged**